File: demos/java-feequote/src/main/java/com/feequote/services/TransferService.java
package com.feequote.services;

import com.feequote.clients.PaymentClient;
import com.feequote.clients.SubmissionReceipt;
import com.feequote.config.Settings;
import com.feequote.core.Pricing;
import com.feequote.models.BreakdownLine;
import com.feequote.models.Quote;
import com.feequote.models.Transfer;
import java.util.List;
import org.springframework.stereotype.Service;

/** Quoting and submission of customer transfers. */
@Service
public class TransferService {

    /**
     * A quoted transfer that has been handed to the gateway.
     *
     * @param quote the price the customer was given
     * @param receipt what the gateway returned
     */
    public record Submission(Quote quote, SubmissionReceipt receipt) {
    }

    private final PaymentClient client;

    public TransferService(PaymentClient client) {
        this.client = client;
    }

    /** Rejects transfers the service cannot price or route. */
    public static void validateTransfer(Transfer transfer) {
        if (transfer.amountCents() <= 0) {
            throw new TransferValidationException("amountCents must be positive");
        }
        if (!Settings.CURRENCIES.contains(transfer.currency())) {
            throw new TransferValidationException("unsupported currency: " + transfer.currency());
        }
        if (!Settings.COUNTRIES.contains(transfer.originCountry())) {
            throw new TransferValidationException("unsupported origin country: " + transfer.originCountry());
        }
        if (!Settings.COUNTRIES.contains(transfer.destinationCountry())) {
            throw new TransferValidationException(
                    "unsupported destination country: " + transfer.destinationCountry());
        }
        if (!Settings.CHANNELS.contains(transfer.channel())) {
            throw new TransferValidationException("unsupported channel: " + transfer.channel());
        }
    }

    /** Prices {@code transfer} and returns the quote with its fee breakdown. */
    public Quote quote(Transfer transfer) {
        validateTransfer(transfer);
        long domesticFee = Pricing.computeFee(transfer.amountCents());
        List<BreakdownLine> breakdown = List.of(new BreakdownLine("domesticFee", domesticFee));
        long feeCents = breakdown.stream().mapToLong(BreakdownLine::amountCents).sum();
        return new Quote(transfer.id(), feeCents, transfer.amountCents() + feeCents, breakdown);
    }

    /** Quotes {@code transfer} and hands it to the gateway. */
    public Submission submit(Transfer transfer) {
        Quote quote = quote(transfer);
        return new Submission(quote, client.submit(transfer));
    }
}

File: demos/python-feequote/feequote/services/transfer_service.py
"""Quoting and submission of customer transfers."""

from __future__ import annotations

from dataclasses import dataclass

from feequote import config
from feequote.clients.payment_client import PaymentClient, SubmissionReceipt
from feequote.core.pricing import compute_fee
from feequote.models import BreakdownLine, Quote, Transfer


class TransferValidationError(ValueError):
    """The transfer cannot be quoted as supplied."""


@dataclass(frozen=True)
class Submission:
    quote: Quote
    receipt: SubmissionReceipt


def validate_transfer(transfer: Transfer) -> None:
    """Reject transfers the service cannot price or route."""
    if transfer.amount_cents <= 0:
        raise TransferValidationError("amount_cents must be positive")
    if transfer.currency not in config.CURRENCIES:
        raise TransferValidationError(f"unsupported currency: {transfer.currency}")
    for label, country in (
        ("origin", transfer.origin_country),
        ("destination", transfer.destination_country),
    ):
        if country not in config.COUNTRIES:
            raise TransferValidationError(f"unsupported {label} country: {country}")
    if transfer.channel not in config.CHANNELS:
        raise TransferValidationError(f"unsupported channel: {transfer.channel}")


class TransferService:
    """Quotes transfers and submits them to the payment gateway."""

    def __init__(self, client: PaymentClient | None = None) -> None:
        self._client = client or PaymentClient()

    def quote(self, transfer: Transfer) -> Quote:
        """Price ``transfer`` and return the quote with its fee breakdown."""
        validate_transfer(transfer)
        domestic_fee = compute_fee(transfer.amount_cents)
        breakdown = (BreakdownLine(label="domestic_fee", amount_cents=domestic_fee),)
        fee_cents = sum(line.amount_cents for line in breakdown)
        return Quote(
            transfer_id=transfer.id,
            fee_cents=fee_cents,
            total_cents=transfer.amount_cents + fee_cents,
            breakdown=breakdown,
        )

    def submit(self, transfer: Transfer) -> Submission:
        """Quote ``transfer`` and hand it to the gateway."""
        quote = self.quote(transfer)
        receipt = self._client.submit(transfer)
        return Submission(quote=quote, receipt=receipt)
